### PR TITLE
DIRECTOR: Transition fixes, palette cycling effects, and debugger

### PR DESCRIPTION
**Pull request by moralrecordings**

### CODE REVIEW

File: engines/director/channel.cpp [hunk 1]
```diff
@@ -107,8 +107,8 @@ Channel::~Channel() {
 
 DirectorPlotData Channel::getPlotData() {
 	DirectorPlotData pd(g_director, _sprite->_spriteType, _sprite->_ink, _sprite->_blend, _sprite->getBackColor(), _sprite->getForeColor());
-	pd.colorWhite = pd.d->_wm->_colorWhite;
-	pd.colorBlack = pd.d->_wm->_colorBlack;
+	pd.colorWhite = 255;
```
Comment (sev-) [inline engines/director/channel.cpp:110]: Why this change? 255 is not guaranteed to be white.

File: engines/director/debugger.cpp [hunk 2]
```diff
@@ -19,12 +19,15 @@
...
 
+#include "common/language.h"
+#include "director/detection.h"
 #include "director/director.h"
 #include "director/lingo/lingo.h"
 #include "director/lingo/lingo-code.h"
 #include "director/lingo/lingo-object.h"
 #include "director/lingo/lingo-codegen.h"
 #include "director/debugger.h"
+#include "game.h"
```
Comment (sev-) [inline engines/director/debugger.cpp:30]: what is `game.h`? All includes must contain relative path

File: engines/director/debugger.cpp [hunk 3]
```diff
@@ -34,24 +37,147 @@ Debugger *g_debugger;
...
+
+bool Debugger::cmdVersion(int argc, const char **argv) {
+	debugPrintf("Director version: %d\n", g_director->getVersion());
+	debugPrintf("Director platform: %s\n", Common::getPlatformCode(g_director->getPlatform()));
+	debugPrintf("Game ID: %s\n", g_director->getGameId());
+	debugPrintf("Game name: %s\n", getDescriptionFromGameId(g_director->getGameId()));
+	debugPrintf("Game variant: %s\n", g_director->getExtra());
+	debugPrintf("Language: %s\n", Common::getLanguageCode(g_director->getLanguage()));
+	debugPrintf("Expected Director version: %d\n", g_director->getDescriptionVersion());
+	debugPrintf("Executable name: %s\n", g_director->getEXEName().c_str());
```
Comment (sev-) [inline engines/director/debugger.cpp:118]: Please print out here md5 and the startup file, as that could help to identify the specific detection entry.

File: engines/director/debugger.cpp [hunk 4]
```diff
@@ -34,24 +37,147 @@ Debugger *g_debugger;
...
+}
+
+bool Debugger::cmdFrame(int argc, const char **argv) {
+	debugPrintf("%s\n", g_director->getLingo()->formatFrame().c_str());
+	debugPrintf("\n");
+	return true;
+}
+
+bool Debugger::cmdBacktrace(int argc, const char **argv) {
+	debugPrintf("%s\n", g_director->getLingo()->formatCallStack(g_director->getLingo()->_pc).c_str());
```
Comment (sev-) [inline engines/director/debugger.cpp:143]: put 

```C++
Lingo *lingo = g_director->getLingo();

debugPrintf("%s\n", lingo->formatCallStack(lingo->_pc).c_str());
```

For better readabilty

File: engines/director/debugger.cpp [hunk 5]
```diff
@@ -34,24 +37,147 @@ Debugger *g_debugger;
...
+
+bool Debugger::cmdFinish(int argc, const char **argv) {
+	_finish = true;
+	_finishCounter = 1;
+	return cmdExit(0, nullptr);
+}
+
 bool Debugger::lingoCommandProcessor(const char *inputOrig) {
-	if (!strcmp(inputOrig, "lingo off")) {
+	if (!strcmp(inputOrig, "exit")) {
```
Comment (sev-) [inline engines/director/debugger.cpp:180]: I am afraid that a number of Director titles could have `exit` as some meaningful keyword. This is why we had `lingo off` in the first place

File: engines/director/lingo/lingo.cpp [hunk 6]
```diff
@@ -743,6 +770,9 @@ void Datum::reset() {
 		case FIELDREF:
 			delete u.cast;
 			break;
+		case MENUREF:
+			delete u.menu;
+			break;
 		default:
 			break;
```
Comment (sev-) [inline engines/director/lingo/lingo.cpp:777]: Let's put a catcher here `warning("Unprocessed REF type %d", type);`

File: engines/director/sprite.cpp [hunk 7]
```diff
@@ -244,13 +244,6 @@ MacShape *Sprite::getShape() {
 		default:
 			break;
 		}
-
-		if (g_director->getVersion() >= 400) {
```
Comment (sev-) [inline engines/director/sprite.cpp:248]: Why is this removed?

File: engines/director/sprite.cpp [hunk 8]
```diff
@@ -265,8 +258,7 @@ uint32 Sprite::getBackColor() {
 
 	switch (_cast->_type) {
 	case kCastText:
-	case kCastButton:
-	case kCastShape: {
```
Comment (sev-) [inline engines/director/sprite.cpp:269]: Why is this removed?

File: engines/director/sprite.cpp [hunk 9]
```diff
@@ -280,8 +272,7 @@ uint32 Sprite::getForeColor() {
...
```
Comment (sev-) [inline engines/director/sprite.cpp:284]: Same question, why is this removed?

File: engines/director/util.h [hunk 10]
```diff
@@ -101,6 +101,25 @@ Common::String utf8ToPrintable(const Common::String &str);
 
 Common::String decodePlatformEncoding(Common::String input);
 
+inline byte lerpByte(byte a, byte b, int alpha, int span) {
```
Comment (sev-) [inline engines/director/util.h:104]: what does `lerp` mean?

File: engines/director/window.cpp [hunk 11]
```diff
@@ -203,7 +203,6 @@ Datum Window::getStageRect() {
 
 void Window::reset() {
 	resize(_composeSurface->w, _composeSurface->h, true);
-	_composeSurface->clear(_stageColor);
```
Comment (sev-) [inline engines/director/window.cpp:206]: Why is this removed? Are we keeping Stage color?

**Review by sev- — CHANGES_REQUESTED** [review]
So, this PR contains the palette changes too...  That is fine, but perhaps we need to close #4241 ?

I put some notes for the patch, needs changes.

File: engines/director/util.h (hunk 10 above)
Comment (moralrecordings) [reply engines/director/util.h:104]: Short for linear interpolation - https://en.wikipedia.org/wiki/Linear_interpolation#History_and_applications

File: engines/director/debugger.cpp [hunk 12]
```diff
@@ -34,22 +36,151 @@ Debugger *g_debugger;
...
+	//debugPrintf(" channels [frameNum] - Shows channel information for a score frame\n");
+	//debugPrintf(" cast - Shows the cast list for the current movie\n");
+	debugPrintf("\n");
+	debugPrintf("Lingo execution:\n");
+	//debugPrintf(" eval [statement] - Evaluates a single Lingo statement\n");
+	debugPrintf(" repl - Switches to a REPL interface for evaluating Lingo code\n");
+	debugPrintf(" backtrace / bt - Prints a backtrace of all stack frames\n");
+	//debugPrintf(" disasm [function] - Lists the bytecode disassembly for a script function\n");
+	debugPrintf(" stack / st - Lists the elements on the stack\n");
+	debugPrintf(" frame / f - Prints the current script frame\n");
```
Comment (rvanlaar) [inline engines/director/debugger.cpp:96]: Please rename `frame`. It conflicts with the common use of frame for a frame in a scene.

File: engines/director/channel.cpp (hunk 1 above)
Comment (sev-) [inline engines/director/channel.cpp:110]: Why this change? 255 is not guaranteed to be white.

**Comment by sev-** [comment]
Thank you!
